src/components/utils/test/posix_thread_test.cc

```diff
@@ -41,287 +43,466 @@ namespace utils_test {
-  if (!result)
-    EXPECT_EQ(test_thread_name, std::string(name));
-  // Stop thread
-  thread->stop();
-  EXPECT_TRUE(
-      thread->start(threads::ThreadOptions(threads::Thread::kMinStackSize)));
-  thread2_id = thread->CurrentId();
-  // Expect the same thread started with the the same name
-  EXPECT_EQ(test_thread_name, std::string(name));
-  EXPECT_EQ(thread1_id, thread2_id);
-  cond_var_.WaitFor(test_lock, 10000);
+
+  {
+    AutoLock test_lock(*threadDelegate);
+    // Create thread
+    ASSERT_NO_THROW(thread = CreateThread(kThreadName, threadDelegate));
+    // Start created thread
+    EXPECT_TRUE(
+        thread->Start(threads::ThreadOptions(threads::Thread::kMinStackSize)));
+    thread1_id = thread->CurrentId();
+    // Rename started thread. Name will be cut to 15 symbols + '\0'
+    // This is the limit in current POSIX thread implementation
+    thread->SetNameForId(thread->ThreadHandle(), kThreadName);
+    // Name must be large enough to keep 16 symbols. Read previous comment
+    char name[kMaxSize];
+    int result = pthread_getname_np(thread->ThreadHandle(), name, sizeof(name));
+    if (!result) {
+      EXPECT_EQ(kThreadName, std::string(name));
+    }
+    threadDelegate->WaitFor(test_lock, 10000);
+    // Stop thread
+    thread->Stop(threads::Thread::kThreadStopTask);
+    EXPECT_TRUE(
+        thread->Start(threads::ThreadOptions(threads::Thread::kMinStackSize)));
+    thread2_id = thread->CurrentId();
+    // Expect the same thread started with the the same name
+    EXPECT_EQ(kThreadName, std::string(name));
+    EXPECT_EQ(thread1_id, thread2_id);
+    threadDelegate->WaitFor(test_lock, 10000);
+  }
+
   EXPECT_TRUE(threadDelegate->check_value());
+
+  delete threadDelegate;
+  EXPECT_EQ(NULL, thread->GetDelegate());
   DeleteThread(thread);
+}
+
+TEST_F(PosixThreadTest, StartThreadWithNullPtrDelegate_ExpectThreadStateNone) {
+  threads::Thread* thread = nullptr;
+  // Create thread
+  ASSERT_NO_THROW(thread = CreateThread(kThreadName, nullptr));
+  EXPECT_TRUE(nullptr != thread);
+  EXPECT_EQ(nullptr, thread->GetDelegate());
+
+  EXPECT_FALSE(thread->Start());
+  EXPECT_FALSE(thread->IsRunning());
+  EXPECT_EQ(threads::Thread::kThreadStateNone, thread->thread_state_);
+
+  DeleteThread(thread);
+}
+
+TEST_F(PosixThreadTest,
+       StartThreadExecutingThreadMain_ExpectThreadStateRunning) {
+  threads::Thread* thread = nullptr;
+  const bool idle_threadMain = true;
+  TestThreadDelegate* threadDelegate = new TestThreadDelegate(idle_threadMain);
+  // Create thread
+  ASSERT_NO_THROW(thread = CreateThread(kThreadName, threadDelegate));
+  EXPECT_TRUE(nullptr != thread);
+
+  {
+    AutoLock auto_lock(*threadDelegate);
+    EXPECT_TRUE(thread->Start());
+    threadDelegate->WaitFor(auto_lock, 5000);
+  }
+
+  EXPECT_TRUE(thread->IsRunning());
+  EXPECT_EQ(threads::Thread::kThreadStateRunning, thread->thread_state_);
+
+  delete threadDelegate;
+  DeleteThread(thread);
```

Comment:
Check via posix_api that thread_handle is not active (may be other UT)

Reply:
@LuxoftAKutsan 
The unfortunate POSIX thread (pthread) does not provide this capability.
We could using the pthread_kill() function with sig is 0, but POSIX says that an attempt to use a thread ID whose lifetime has ended produces undefined behavior.
Or pthread_join, but Joining with a thread that has previously been joined results in undefined behavior